src/snarks/groth16/mpc/phase2.hpp

```diff
@@ -0,0 +1,309 @@
+/// proof-of-knowledge, consisting of some point $s$ in G1, chosen randomly, $s
+/// * \delta$, $r$ generated deterministically by hashing $s$ and $s * \delta$
+/// into G2, and $r * \delta$.
+///
+/// Similarly to the "Phase2" library (https://github.com/ebfull/phase2), we
+/// track both the new value of delta (after applying this contribution),
+/// and the transaction digest before applying, in order that we can
+/// efficiently check the transcript using just the public keys.
+template<typename ppT> class srs_mpc_phase2_publickey
+{
+public:
+    srs_mpc_hash_t transcript_digest;
+    libff::G1<ppT> new_delta_g1;
+    libff::G1<ppT> s_g1;
+    libff::G1<ppT> s_delta_j_g1;
+    libff::G2<ppT> r_delta_j_g2;
+
+    srs_mpc_phase2_publickey(
+        const srs_mpc_hash_t transcript_digest,
+        const libff::G1<ppT> &new_delta_g1,
+        const libff::G1<ppT> &s_g1,
+        const libff::G1<ppT> &s_delta_j_g1,
+        const libff::G2<ppT> &r_delta_j_g2);
+
+    bool operator==(const srs_mpc_phase2_publickey<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    static srs_mpc_phase2_publickey<ppT> read(std::istream &in);
+    void compute_digest(srs_mpc_hash_t out_digest) const;
+};
+
+/// Challenge given to a participant in Phase2 of the SRS generation MPC.
+template<typename ppT> class srs_mpc_phase2_challenge
+{
+public:
+    srs_mpc_hash_t transcript_digest;
+    srs_mpc_phase2_accumulator<ppT> accumulator;
+
+    srs_mpc_phase2_challenge(
+        const srs_mpc_hash_t transcript_digest,
+        srs_mpc_phase2_accumulator<ppT> &&accumulator);
+
+    bool operator==(const srs_mpc_phase2_challenge<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    static srs_mpc_phase2_challenge<ppT> read(std::istream &in);
+};
+
+/// Reponse produced by participant in Phase2 of the SRS generation MPC.
+template<typename ppT> class srs_mpc_phase2_response
+{
+public:
+    srs_mpc_phase2_accumulator<ppT> new_accumulator;
+    srs_mpc_phase2_publickey<ppT> publickey;
+
+    srs_mpc_phase2_response(
+        srs_mpc_phase2_accumulator<ppT> &&new_accumulator,
+        srs_mpc_phase2_publickey<ppT> &&publickey);
+
+    bool operator==(const srs_mpc_phase2_response<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    static srs_mpc_phase2_response<ppT> read(std::istream &in);
+};
+
+// Phase2 functions
+
+template<mp_size_t n, const libff::bigint<n> &modulus>
+void srs_mpc_digest_to_fp(
+    const srs_mpc_hash_t transcript_digest,
+    libff::Fp_model<n, modulus> &out_fr);
+
+template<typename ppT>
+libff::G2<ppT> srs_mpc_digest_to_g2(const srs_mpc_hash_t digest);
+
+/// Given the output from the linear combination of the L1 layer of the SRS
+/// circuit, compute the starting parameters for Phase 2 (the MPC for C2
+/// layer).  See "Initialization" in section 7.3 of [BoweGM17].
```

Comment:
Spacing after `.` (applicable to other places below)

Reply:
Removed space